Sources/Display.php

```diff
@@ -842,45 +842,163 @@ function Display()
-			'max' => $limit,
-		)
-	);
+	// Check if we can use seek
+	if (isset($_SESSION['page_topic']) && isset($_SESSION['page_next_start']) && $_SESSION['page_topic'] == $topic && $_SESSION['page_next_start'] == $start)
+	{	// yes we call the next page
+		$start_char = 'M'; 
+		$page_id = $_SESSION['page_last_id'];
+	}
+	elseif (isset($_SESSION['page_topic']) && isset($_SESSION['page_before_start']) && $_SESSION['page_topic'] == $topic && $_SESSION['page_before_start'] == $start)
+	{	// yes we go backward
+		$start_char = 'L';
+		$page_id = $_SESSION['page_first_id'];
+	}
+	elseif (isset($_SESSION['page_topic']) && isset($_SESSION['page_current_start']) && $_SESSION['page_topic'] == $topic && $_SESSION['page_current_start'] == $start)
+	{	// refresh of current page
+		$start_char = 'C';
+		$page_id = $_SESSION['page_first_id'];
+	}
+	elseif ($start == 0) //special case start page
+	{	// special case start page
+		$start_char = 'C';
+		$page_id = $context['topicinfo']['id_first_msg'];
+	}
+	else
+		$start_char = null;
 
+	$limit = $context['messages_per_page'];
+	
 	$messages = array();
 	$all_posters = array();
-	while ($row = $smcFunc['db_fetch_assoc']($request))
+	
+	if (isset($start_char))
 	{
-		if (!empty($row['id_member']))
-			$all_posters[$row['id_msg']] = $row['id_member'];
-		$messages[] = $row['id_msg'];
+		$firstIndex = 0;
+
+		if ($start_char === 'M' or $start_char === 'S' or $start_char === 'C')
+		{
+			$ascending = true;
+			$page_operator = '>=';
+		}
+		else
+		{
+			$ascending = false;
+			$page_operator = '<=';
+		}
+
+		if ($start_char === 'C')
+			$limit_seek = $limit;
+		else
+			$limit_seek  = $limit + 1;
+
+		$request = $smcFunc['db_query']('', '
+			SELECT id_msg, id_member, approved
+			FROM {db_prefix}messages
+			WHERE id_topic = {int:current_topic} 
+			AND id_msg '. $page_operator . ' {int:page_id}'. (!$modSettings['postmod_active'] || $approve_posts ? '' : '
+			AND (approved = {int:is_approved}' . ($user_info['is_guest'] ? '' : ' OR id_member = {int:current_member}') . ')') . '
+			ORDER BY id_msg ' . ($ascending ? '' : 'DESC') . ($context['messages_per_page'] == -1 ? '' : '
+			LIMIT {int:limit}'),
+			array(
+				'current_member' => $user_info['id'],
+				'current_topic' => $topic,
+				'is_approved' => 1,
+				'blank_id_member' => 0,
+				'limit' => $limit_seek,
+				'page_id' => $page_id,
+			)
+		);
+
+		$found_msg = false;
+
+		// Fallback
+		if ($smcFunc['db_num_rows']($request) < 1)
+			unset($start_char);
+		else
+		{
+			while ($row = $smcFunc['db_fetch_assoc']($request))
+			{
+				if ($row['id_msg'] != $page_id || $start_char === 'C')
+				{
+					if (!empty($row['id_member']))
+						$all_posters[$row['id_msg']] = $row['id_member'];
+					$messages[] = $row['id_msg'];
+
+					if ($start_char === 'C' && $row['id_msg'] == $page_id) // find the start message of current page
+						$found_msg = true;
+				}
+				else
+					$found_msg = true; // find the start message of next or before page
```

Comment:
When I read correctly (for me is this version harder to  read)
```
if ($row['id_msg'] == $page_id)
	$found_msg = true;

if ($start_char === 'C' || !$found_msg)
{
	if (!empty($row['id_member']))
		$all_posters[$row['id_msg']] = $row['id_member'];

	$messages[] = $row['id_msg'];
}
```
Doesn't skip the 16th dataset for the L and M case.
```
if ($row['id_msg'] == $page_id)
	$found_msg = true;

if ($start_char === 'C' || $row['id_msg'] != $page_id)
{
	if (!empty($row['id_member']))
		$all_posters[$row['id_msg']] = $row['id_member'];

	$messages[] = $row['id_msg'];
}
```
This would be logically equivalent in my eyes.